### PR TITLE
[FLINK-33559] Externalize Kafka Python connector code

### PR DESCRIPTION
**Pull request by pvary**
Moves the `kafka.py` and the `kafka_test.py` code from the main Flink repository the connector repository.

When running `mvn clean install` downloads the testing infra scripts from the main Flink repo (`flink-python/dev/build-wheels.sh`, `flink-python/dev/install_command.sh`, `flink-python/dev/lint-python.sh`)

After that the tests could be run by the following commands:
```
cd flink-python
chmod a+x dev/* 
./dev/lint-python.sh -e mypy,sphinx
```

**Test changes**
The `kafka_test.py` had to be modified by adding [_check_latest_offsets_initializer](https://github.com/apache/flink-connector-kafka/pull/69/files#diff-c2fba8348f66807150d91656172726fb68bdfce68bfb067cc0c5a25da27d542dR359) because of [FLINK-28303](https://issues.apache.org/jira/browse/FLINK-28303) #52 

**Gateway server patch**
https://github.com/apache/flink/pull/23843 is needed in the long run. As a temporary solution the `pyflink_gateway_server.py` is duplicated and the changes applied here. During the `integration_test.sh` we replace the `pyflink_gateway_server.py` coming from the apache-flink package with this patched one.
This temporary solution should be removed after migrating to the apache-flink 1.19. The 2 plces:
- [integration_test.sh](https://github.com/apache/flink-connector-kafka/pull/69/files#diff-1cf9b67f6aa518c8765f9b15b6cce647623fcb58da25078b385a4f624675fe87R48-R51)
- [pyflink_gateway_server.py](https://github.com/apache/flink-connector-kafka/pull/69/files#diff-61da3b786b6f90cda5f5be9e82446adc94794d1c9fb16b5ec9b1b3636ef881a5)

**CI patch**
To run the python tests on CI https://github.com/apache/flink-connector-shared-utils/pull/27 is needed. After the shared-utils change is merged, [push_pr.yml](https://github.com/apache/flink-connector-kafka/pull/69/files#diff-3df4df4ee2f15512e27ab35c6b31295f36d13ebd4823fe108026221c562da861R37) should be changed from
```
uses: pvary/flink-connector-shared-utils/.github/workflows/python_ci.yml@ci_utils
```
to
```
uses: apache/flink-connector-shared-utils/.github/workflows/python_ci.yml@ci_utils
```
In the meantime, here is the result of the [CI run](https://github.com/pvary/flink-connector-kafka/actions/runs/7047273599)

### CODE REVIEW

File: .github/workflows/push_pr.yml [hunk 1]
```diff
@@ -29,3 +29,11 @@ jobs:
     uses: apache/flink-connector-shared-utils/.github/workflows/ci.yml@ci_utils
     with:
       flink_version: ${{ matrix.flink }}
+    
+  python_test:
+    strategy:
+      matrix:
+        flink: [ 1.17.1, 1.18.0 ]
+    uses: pvary/flink-connector-shared-utils/.github/workflows/python_ci.yml@ci_utils
```
Comment (pvary) [inline .github/workflows/push_pr.yml:37]: This should be changed to `apache/flink-connector-shared-utils/.github/workflows/python_ci.yml@ci_utils` once https://github.com/apache/flink-connector-shared-utils/pull/27 merged

File: flink-python/setup.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import glob
+import io
+import os
+import sys
+
+from setuptools import setup
+from shutil import copy, rmtree
+from xml.etree import ElementTree as ET
+
+PACKAGE_NAME = 'apache-flink-connectors-kafka'
```
Comment (dianfu) [inline flink-python/setup.py:29]: Do you think it makes sense to rename it to `apache-flink-connector-kafka`, that's removing the character `s`. My motivation is to keep it consistent with the repository name.

File: flink-python/tox.ini [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+[tox]
+# tox (https://tox.readthedocs.io/) is a tool for running tests
+# in multiple virtualenvs. This configuration file will run the
+# test suite on all supported python versions.
+# new environments will be excluded by default unless explicitly added to envlist.
+envlist = {py38, py39, py310}-cython
```
Comment (dianfu) [inline flink-python/tox.ini:24]: py311 is missing

File: flink-python/pom.xml [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    <parent>
+        <groupId>org.apache.flink</groupId>
+        <artifactId>flink-connector-kafka-parent</artifactId>
+        <version>3.1-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>flink-connector-kafka-python</artifactId>
+    <name>Flink : Connectors : SQL : Kafka : Python</name>
+
+    <packaging>jar</packaging>
```
Comment (dianfu) [inline flink-python/pom.xml:35]: Actually there is no JAR package here. What about changing it to pom?

File: flink-python/setup.py (hunk 2 above)
Comment (pvary) [reply flink-python/setup.py]: Done

File: flink-python/pom.xml (hunk 4 above)
Comment (pvary) [reply flink-python/pom.xml]: Done

File: flink-python/setup.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    long_description=long_description,
+    long_description_content_type='text/plain',
+    zip_safe=False,
+    classifiers=[
+        'Development Status :: 5 - Production/Stable',
+        'License :: OSI Approved :: Apache Software License',
+        'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
+        'Programming Language :: Python :: 3.11']
```
Comment (gaborgsomogyi) [inline flink-python/setup.py:155]: I think we're not yet supporting 3.11 here but 3.7, right?

**Review by gaborgsomogyi — APPROVED** [review]
LGTM. If there are no further comments then intended to merge it next Monday.

**Review by dianfu — APPROVED** [review]

**Pull request merged**